### PR TITLE
Non-intrusive API for calling mypy directly from another python application.

### PR DESCRIPTION
**Pull request by JdeH**
After an earlier attempt (#2114), which was by @gvanrossum deemed to complex and restrictive in the current development stage of mypy, as promised here's another shot at it, indeed simpler and closer to the command line. This API uses mypy completely as a black box, not posing any demands to its internals.

Just `import mypy.api` and call the `run` function, with exactly the params that would normally have been passed on the command line, and it will run mypy without starting an new interpreter instance or going through the OS.

The `run` function will return a tuple of strings, the first one containing the output normally routed to std_out, the second one containing the output normally routed to std_err.

While mypy primarily is meant as a command line tool, my usecase was that I wanted to integrate it into the [Transcrypt](https://github.com/QQuick/Transcrypt) Python to JS compiler without the overhead of going through the OS. I saw a few similar requests in the issues. Since this API doesn't have any interaction with the guts of mypy, it will cover these use cases without adding complexity or restrictions.

Since it follows the command line conventions closely, the command line docs and the comments in the source code provide enough info to cover the use of this API. Still, should any extra docs or usage examples be required, I'll happily provide that.

Kind regards
Jacques de Hooge

### CODE REVIEW

File: mypy/api.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+'''
```
Comment (JukkaL) [inline mypy/api.py]: Style nit: Use triple double quotes and the summary on the first line (example: `"""Single-line summary. ...`) for docstrings.

File: mypy/api.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    print('\nError report:\n')
+    print(result[1])  # stderr
+'''
+
+import sys
+from io import StringIO
+from mypy.main import main
+
+
+def run(params):
```
Comment (JukkaL) [inline mypy/api.py]: Add type annotation.

File: mypy/api.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    try:
+        main(None)
+    except:
+        pass
+
+    sys.stdout = old_stdout
+    sys.stderr = old_stderr
+
+    return(new_stdout.getvalue(), new_stderr.getvalue())
```
Comment (JukkaL) [inline mypy/api.py]: Style nits: Add space after `return`. No need to add parentheses around the return value expression.

**Comment by JukkaL** [comment]
This looks reasonable to me -- it's simple and not intrusive at all. I had just few minor nits.

**Comment by JdeH** [comment]
Will adapt the nits and resubmit

**Comment by wtpayne** [comment]
This is very timely for me. I am looking forward to making use of it once it is merged.

**Comment by gvanrossum** [comment]
Hey @JdeH, are you still planning to push a new version responding to Jukka's code review? It sounded like it's almost there.

**Comment by JdeH** [comment]
Hi @gvanrossum , I've completed the changes atwo weeks ago, but the problem is I can't get the newest version of mypy properly installed to test it. I felt a bit stupid and tried to find out what to do, but after several hours I didn't make any progress. The error report during installation for Python 3.5, on Windows (and almost the same on Linux) is:
```

Collecting dbutils (from sqlbean>=0.490->mypy)
  Using cached DBUtils-1.1.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\info_000\AppData\Local\Temp\pip-build-s58mr4b5\dbutils\setup.py", line 11, in <module>
        raise ImportError('Python %d.%d is not supported by DBUtils.' % py_version)
    ImportError: Python 3.5 is not supported by DBUtils.
    
    ----------------------------------------
```

Since the `api.py` doesn't break anything and someone seems to need it, I'll commit it as is. If there turn out to be issues with it, I'll take a look, maybe with some help of another contributer.

Also any hint on how to properly install mypy is welcome. I've spent quite some time with the docs, but must be missing something obvious.

**Comment by gvanrossum** [comment]
OK, I'll merge this now. Thanks!

**Comment by ohadraz** [comment]
documentation and/or usage for this?

**Comment by dandavison** [comment]
> documentation and/or usage for this?

https://mypy.readthedocs.io/en/stable/extending_mypy.html#integrating-mypy-into-another-python-application

**Pull request merged**